src/shell/input.c

```diff
@@ -0,0 +1,710 @@
+}
+
+static void shell_task_input_cleanup (struct shell_task_input *tp)
+{
+    shell_task_input_kvs_cleanup (&(tp->input_kvs));
+}
+
+static void shell_input_type_file_cleanup (struct shell_input_type_file *fp)
+{
+    close (fp->fd);
+    flux_watcher_destroy (fp->w);
+    free (fp->rankstr);
+}
+
+void shell_input_destroy (struct shell_input *in)
+{
+    if (in) {
+        int saved_errno = errno;
+        int i;
+        shell_input_type_file_cleanup (&(in->stdin_file));
+        for (i = 0; i < in->ntasks; i++)
+            shell_task_input_cleanup (&(in->task_inputs[i]));
+        free (in->task_inputs);
+        free (in);
+        errno = saved_errno;
+    }
+}
+
+static void shell_input_put_kvs_completion (flux_future_t *f, void *arg)
+{
+    struct shell_input *in = arg;
+
+    if (flux_future_get (f, NULL) < 0)
+        /* failng to write stdin to input is a fatal error.  Should be
+         * cleaner in future. Issue #2378 */
+        log_err_exit ("shell_input_put_kvs");
+    flux_future_destroy (f);
+
+    if (flux_shell_remove_completion_ref (in->shell, "input.kvs") < 0)
+        log_err ("flux_shell_remove_completion_ref");
+}
+
+static int shell_input_put_kvs (struct shell_input *in, json_t *context)
+{
+    flux_kvs_txn_t *txn = NULL;
+    flux_future_t *f = NULL;
+    json_t *entry = NULL;
+    char *entrystr = NULL;
+    int saved_errno;
+    int rc = -1;
+
+    if (!(entry = eventlog_entry_pack (0.0, "data", "O", context)))
+        goto error;
+    if (!(entrystr = eventlog_entry_encode (entry)))
+        goto error;
+    if (!(txn = flux_kvs_txn_create ()))
+        goto error;
+    if (flux_kvs_txn_put (txn, FLUX_KVS_APPEND, "input", entrystr) < 0)
+        goto error;
+    if (!(f = flux_kvs_commit (in->shell->h, NULL, 0, txn)))
+        goto error;
+    if (flux_future_then (f, -1, shell_input_put_kvs_completion, in) < 0)
+        goto error;
+    if (flux_shell_add_completion_ref (in->shell, "input.kvs") < 0) {
+        log_err ("flux_shell_remove_completion_ref");
+        goto error;
+    }
+    /* f memory responsibility of shell_input_put_kvs_completion()
+     * callback */
+    f = NULL;
+    rc = 0;
+ error:
+    saved_errno = errno;
+    flux_kvs_txn_destroy (txn);
+    free (entrystr);
+    json_decref (entry);
+    flux_future_destroy (f);
+    errno = saved_errno;
+    return rc;
+}
+
+/* Convert 'iodecode' object to an valid RFC 24 data event.
+ * N.B. the iodecode object is a valid "context" for the event.
+ */
+static void shell_input_stdin_cb (flux_t *h,
+                                  flux_msg_handler_t *mh,
+                                  const flux_msg_t *msg,
+                                  void *arg)
+{
+    struct shell_input *in = arg;
+    bool eof = false;
+    json_t *o;
+
+    if (flux_request_unpack (msg, NULL, "o", &o) < 0)
+        goto error;
+    if (iodecode (o, NULL, NULL, NULL, NULL, &eof) < 0)
+        goto error;
+    if (shell_svc_allowed (in->shell->svc, msg) < 0)
```

Comment:
Should `shell_svc_allowed()` be called before unpacking the message and decoding input object? O/w malicious user could send a malformed message or payload in an attempt to compromise the shell.

Reply:
good catch, this is done in the output side as well and needs to be corrected in both.

Reply:
Oh good catch!